### DISCUSSION TITLE
DuckDB export to Pandas - is it a zero-copy process?

### DISCUSSION THREAD

**Discussion by match-gabeflores** [duckdb/duckdb#9911] [Q&A] [answered]
Polars can convert to Pandas and according to them, it's a zero-copy process, because it uses pyarrow extension array option. Does DuckDB do this to make it a zero-copy process? Any gotchas?

If so, I think it'd be good to document here :) 
https://duckdb.org/docs/archive/0.9.2/guides/python/export_pandas

https://github.com/pola-rs/polars/blob/py-0.19.19/py-polars/polars/dataframe/frame.py#L3553

**Comment by pdet** [comment, marked as answer]
We currently only support conversion to/from Pandas using the old backend, and the pyarrow backend is not yet supported. 

This means that some columns will be zero-copied if they can be expressed as a NumPy array.

**Reply by match-gabeflores** [reply]
Thanks for the answer! Are there plans to use the pyarrow backend? Do you advise I submit a feature request?

**Reply by pdet** [reply]
Absolutely! I began working on it some time ago, but currently, I don't have the time to finish it off. If you're interested in taking on this task, I would be happy to review your PR.

Here is my branch with some very initial work on it:
https://github.com/pdet/duckdb/tree/pandas_warning

edit: Oh I read that you wanted to do a pull request

**Reply by match-gabeflores** [reply]
I wish I could help with a PR but it's not my expertise!

**Reply by zmbc** [reply]
Apologies if this is obvious, but what does it mean that "some columns will be zero-copied if they can be expressed as a NumPy array"?

My understanding was that _all_ DuckDB columns were stored in arrow format, so converting anything to NumPy would not be a zero-copy process.

**Reply by Tishj** [reply]
DuckDB does not use arrow internally

And I dont think this holds true anymore either:
> some columns will be zero-copied if they can be expressed as a NumPy array

Never the less, we are looking to add the pyarrow backend for our pandas export somewhere down the line
